metricbeat/module/prometheus/collector/collector.go

```diff
@@ -177,9 +189,48 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
 
 	// Report events
 	for _, e := range eventList {
-		isOpen := reporter.Event(mb.Event{
-			RootFields: mapstr.M{m.namespace: e},
-		})
+		event := mb.Event{RootFields: mapstr.M{m.namespace: e}}
+
+		if m.metricsCount {
+			// In x-pack prometheus module, the metrics are nested under the "prometheus" key directly.
+			// whereas in non-x-pack prometheus module, the metrics are nested under the "prometheus.metrics" key.
+			// Also, it is important that we do not just increment by 1 for each metric because histograms and summaries are special.
+			// For example, if you notice histogram's implementation in data.go, then you'd notice single PromEvent holds 2 metrics. Here:
+			//
+			//	 PromEvent{
+			//	 	Data: mapstr.M{
+			//	 		"metrics": mapstr.M{
+			//	 			name + "_sum":   histogram.GetSampleSum(),
+			//	 			name + "_count": histogram.GetSampleCount(),
+			//	 		},
+			//	 	},
+			//	 	Labels: labels,
+			// 	}
+			//
+			// Here, name +  "_sum" and name + "_count" are the 2 metrics.
+			//
+			// So, len(v) will be 2 in the above example.
+			// Similarly, it will happen for x-pack prometheus module too. Please see
+			// the unit tests for the same.
+			switch m.xPack {
+			case true:
+				// As, metrics are nested under the "prometheus" key in case of x-pack,
+				// labels is also nested under the "prometheus" key. So, we need to
+				// make sure we subtract 1 in case the e["labels"]
+				// also exists.
+				if _, hasLabels := e["labels"].(mapstr.M); hasLabels {
+					event.RootFields.Put("metrics_count", len(e)-1)
+				} else {
+					event.RootFields.Put("metrics_count", len(e))
+				}
+			default:
+				if v, ok := e["metrics"].(mapstr.M); ok {
+					event.RootFields.Put("metrics_count", len(v))
+				}
+			}
```

Comment:
I'm thinking if we need this additional variable xPack at all. Since we have access to the metricset object, we could simply
```suggestion
			switch m.promEventsGen.(type) { {
			case *promEventGenerator: // x-pack
			          if v, ok := e["metrics"].(mapstr.M); ok {
					event.RootFields.Put("metrics_count", len(v))
				 }
			default: // non-xpack
				// As, metrics are nested under the "prometheus" key in case of x-pack,
				// labels is also nested under the "prometheus" key. So, we need to
				// make sure we subtract 1 in case the e["labels"]
				// also exists.
				if _, hasLabels := e["labels"].(mapstr.M); hasLabels {
					event.RootFields.Put("metrics_count", len(e)-1)
				} else {
					event.RootFields.Put("metrics_count", len(e))
				}
			}
```

Reply:
Yes, my first draft change was this only. 

But this may seem unclear to some future readers of the code, so I added the xPack so that it is clear, and they can also use the xPack variable in case they need it.

What do you think?

Reply:
IMO, since you already added a note about the generator type while asserting it, it should be enough the readers to know why it is the way it is. Besides, metricset is accessible from both collector as well as remote write. So, we should be able to use no matter where it's needed. 

That said, I am all in for keeping it this way too. My question was more regarding the use of an extra field, which seemed redundant to me.

Reply:
I did take a look at this and I recall another rationale behind introducing a new boolean variable. The primary goal was to optimize performance by eliminating the need for repeated type-checking in every `Fetch` operation (happens at set intervals). By performing the type assertion once during initialization and storing the result in a boolean field, we avoid the overhead of type-checking in each `Fetch` call. So it replaces frequent type assertions with a simple boolean check.

So keeping it as it is?

Reply:
Sounds good. Resolving the conversation in favour of keeping it unchanged.